src/hugr/view.rs

```diff
@@ -77,35 +97,66 @@ pub trait HugrView {
 
     /// Number of inputs to a node.
     /// Shorthand for [`num_ports`][HugrView::num_ports]`(node, Direction::Incoming)`.
-    fn num_inputs(&self, node: Node) -> usize;
+    #[inline]
+    fn num_inputs(&self, node: Node) -> usize {
+        self.num_ports(node, Direction::Incoming)
+    }
 
     /// Number of outputs from a node.
     /// Shorthand for [`num_ports`][HugrView::num_ports]`(node, Direction::Outgoing)`.
-    fn num_outputs(&self, node: Node) -> usize;
+    #[inline]
+    fn num_outputs(&self, node: Node) -> usize {
+        self.num_ports(node, Direction::Outgoing)
+    }
 
     /// Return iterator over children of node.
-    fn children(&self, node: Node) -> Children<'_>;
+    fn children(&self, node: Node) -> Self::Children<'_>;
 
     /// Iterates over neighbour nodes in the given direction.
     /// May contain duplicates if the graph has multiple links between nodes.
-    fn neighbours(&self, node: Node, dir: Direction) -> Neighbours<'_>;
+    fn neighbours(&self, node: Node, dir: Direction) -> Self::Neighbours<'_>;
 
     /// Iterates over the input neighbours of the `node`.
     /// Shorthand for [`neighbours`][HugrView::neighbours]`(node, Direction::Incoming)`.
-    fn input_neighbours(&self, node: Node) -> Neighbours<'_>;
+    #[inline]
+    fn input_neighbours(&self, node: Node) -> Self::Neighbours<'_> {
+        self.neighbours(node, Direction::Incoming)
+    }
 
     /// Iterates over the output neighbours of the `node`.
     /// Shorthand for [`neighbours`][HugrView::neighbours]`(node, Direction::Outgoing)`.
-    fn output_neighbours(&self, node: Node) -> Neighbours<'_>;
+    #[inline]
+    fn output_neighbours(&self, node: Node) -> Self::Neighbours<'_> {
+        self.neighbours(node, Direction::Outgoing)
+    }
 
     /// Iterates over the input and output neighbours of the `node` in sequence.
-    fn all_neighbours(&self, node: Node) -> Neighbours<'_>;
+    fn all_neighbours(&self, node: Node) -> Self::Neighbours<'_>;
 }
 
 impl<T> HugrView for T
 where
     T: AsRef<Hugr>,
 {
+    /// An Iterator over the nodes in a Hugr(View)
+    type Nodes<'a> = MapInto<multiportgraph::Nodes<'a>, Node> where Self: 'a;
+
+    /// An Iterator over (some or all) ports of a node
+    type NodePorts<'a> = MapInto<portgraph::portgraph::NodePortOffsets, Port> where Self: 'a;
+
+    /// An Iterator over the children of a node
+    type Children<'a> = MapInto<portgraph::hierarchy::Children<'a>, Node> where Self: 'a;
+
+    /// An Iterator over (some or all) the nodes neighbouring a node
+    type Neighbours<'a> = MapInto<multiportgraph::Neighbours<'a>, Node> where Self: 'a;
+
+    /// Iterator over the children of a node
+    type PortLinks<'a> = MapCtx<
+        WithCtx<multiportgraph::PortLinks<'a>, &'a Hugr>,
+        fn((multiportgraph::SubportIndex, multiportgraph::SubportIndex), &&'a Hugr) -> (Node, Port),
```

Comment:
Oh, blimey. Are you sure this is better than PortLinks? The old PortLinks seems a lot easier to read and understand IMHO, and you could keep that without changing the rest of the PR...

Reply:
I guess this might get us the "win" of being a DoubleEndedIterator for free? Which is something, I admit. Note that HugrView doesn't specify that the iterators are double-ended, though - only this particular implementation does - do we want it to?

Reply:
I pushed a way cleaner version instead. What do you think about it?
(`MapWithCtx` will be in the funky library at some point :) )

Reply:
> I guess this might get us the "win" of being a DoubleEndedIterator for free?

The main goal is dropping the hand-written iterators, which just bloat the code and always end up missing some trait implementation.

This is not as big of a problem here as it is in portgraph (which has 800+ lines of trivial iterator impls), but the future region views will also require the same iterator-wrapping code.